larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -312,13 +329,24 @@ void OptimizeLCE::runOnFunction() {
 
 // Creates an instance of the TensorFlow dialect OptimizeLCE pass.
 std::unique_ptr<OperationPass<FuncOp>> CreateOptimizeLCEPass(
-    bool experimental_enable_bitpacked_activations) {
-  return std::make_unique<OptimizeLCE>(
-      experimental_enable_bitpacked_activations);
+    const LCETarget target,
+    const bool experimental_enable_bitpacked_activations) {
+  if (target == LCETarget::ARM) {
+    return std::make_unique<OptimizeLCE<LCETarget::ARM>>(
+        experimental_enable_bitpacked_activations);
+  } else {
+    return std::make_unique<OptimizeLCE<LCETarget::XCORE>>(
+        experimental_enable_bitpacked_activations);
+  }
 }
 
-static PassRegistration<OptimizeLCE> pass(
-    "tfl-optimize-lce", "Optimize within the TensorFlow Lite dialect");
+static PassRegistration<OptimizeLCE<LCETarget::ARM>> pass_target_arm(
+    "tfl-optimize-lce-target-arm",
+    "Optimize within the TensorFlow Lite dialect, for ARM targets");
+
+static PassRegistration<OptimizeLCE<LCETarget::XCORE>> pass_target_xcore(
+    "tfl-optimize-lce-target-xcore",
+    "Optimize within the TensorFlow Lite dialect, for XCORE targets");
```

Comment:
This is how two target-specific passes are registered for the purposes of the file-check tests.

The reason that `target` has to be a template argument of the `OptimizeLCE` struct is because when registering MLIR passes, the key that is used is derived from the type of the pass. This means it's not possible to register more than one pass of the same type! I discoved this after lots of trial-and-error. The relevant MLIR function is [here](https://github.com/llvm/llvm-project/blob/62ec4ac90738a5f2d209ed28c822223e58aaaeb7/mlir/lib/Pass/PassRegistry.cpp#L105-L111) - note that if the TODO there had been completed this wouldn't be an issue.

Reply:
I don't think I fully understand why we need two dedicated passes. Why can't we handle this through a pass option similar to [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/lower_static_tensor_list.cc#L94-L99) and [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/tests/lower-static-tensor-list.mlir#L1) or [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/quantization/import_quant_stats_pass.cc#L42-L45) and [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/quantization/tests/import_quant_stats.mlir#L1)?

Reply:
There is a very answer to this, and it's that I wasn't aware this was an option :)

Thanks for finding this, I guess we should have been using this already for the `experimental_enable_bitpacked_activations` option of the optimise pass.

Reply:
Sounds good :)

> I guess we should have been using this already for the experimental_enable_bitpacked_activations option of the optimise pass.

Yes, although I don't think there it really mattered since we never really tested it with the flag disabled.

Reply:
I've just pushed an updated commit using these flags.